### ISSUE TITLE
CUDA_ERROR_ILLEGAL_ADDRESS encountered upon certain condition

### ISSUE THREAD

**Issue by zxytim** [google/jax#7816]
## Code

```python
from typing import Any
import argparse
import functools

import jax
import jax.numpy as jnp

import optax

from flax.training import train_state
from flax import linen as nn


class ResBlock(nn.Module):
    channels: int
    stride: int
    norm: Any

    @nn.compact
    def __call__(self, x):
        shortcut = x
        act = nn.leaky_relu

        c, s = self.channels, self.stride

        x = act(self.norm()(x))
        x = nn.Conv(c // 4, (1, 1))(x)

        x = act(self.norm()(x))
        x = nn.Conv(c // 4, (3, 3), (s, s))(x)

        x = nn.Conv(c, (1, 1))(x)

        if shortcut.shape != x.shape:
            x = self.norm()(x)
            shortcut = nn.Conv(c, (1, 1), (s, s), name='conv_proj')(shortcut)
        return shortcut + x


class Network(nn.Module):
    @nn.compact
    def __call__(self, x, train: bool=True):
        x = nn.Conv(
            16, (3, 3), (2, 2), name='conv_init'
        )(x)
        x = nn.leaky_relu(nn.GroupNorm(16)(x))
        x = nn.max_pool(x, (3, 3), (2, 2), padding='SAME')

        stage_sizes = [2, 2, 3, 2]

        for i, block_size in enumerate(stage_sizes):
            for j in range(block_size):
                s = 2 if i > 0 and j == 0 else 1
                c = 16 * 2**i * 4
                # print(i, j, c, s, x.shape)
                x = ResBlock(channels=c, stride=s,
                             norm=nn.GroupNorm
                             if i > 0
                             else functools.partial(nn.GroupNorm, 16)
                             )(x)
                # print(x.shape)
        x = jnp.mean(x, axis=(1, 2))
        x = nn.Dense(1000)(x)
        return x


def xent_loss_one_hot(logits, labels):
    return jnp.mean(optax.softmax_cross_entropy(
        logits, jax.nn.one_hot(labels, 1000)))


def compute_metrics(logits, labels):
    loss = xent_loss_one_hot(logits, labels)
    accuracy = jnp.mean(jnp.argmax(logits, -1) == labels)
    metrics = {
        'loss': loss,
        'accuracy': accuracy
    }
    return metrics


def train_step(state, x, y):
    def loss_fn(params):
        logits = Network().apply({'params': params}, x)
        loss = xent_loss_one_hot(logits, y)
        return loss, logits

    grad_fn = jax.value_and_grad(loss_fn, has_aux=True)
    (_, logits), grads = grad_fn(state.params)
    state = state.apply_gradients(grads=grads)
    metrics = compute_metrics(logits, y)

    return state, metrics


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('--batch-size', type=int, required=True)
    parser.add_argument('--jit', action='store_true')
    args = parser.parse_args() 

    # Batch size 8 --> works
    # Batch size 16 --> CUDA_ERROR_ILLEGAL_ADDRESS 
    batch_size = args.batch_size

    rng = jax.random.PRNGKey(42)
    rng, init_rng = jax.random.split(rng)

    net = Network()
    params = net.init(rng, jnp.ones((batch_size, 224, 224, 3)))['params']

    tx = optax.adamw(learning_rate=1e-3, weight_decay=5e-4)
    state = train_state.TrainState.create(
        apply_fn=net.apply, params=params, tx=tx)
 
    # IMPORTANT: we use "batch_size - 1" here
    x, y = jnp.ones((batch_size - 1, 224, 224, 3)), jnp.zeros(batch_size - 1)

    if args.jit:
        train = jax.jit(train_step)
    else:
        train = train_step
        
    state, metrics = train(state, x, y)
    print(state.params['conv_init']['bias'])

    print('Have a nice day!')


if __name__ == '__main__':
    main()
```

## Symptoms
`CUDA_ERROR_ILLEGAL_ADDRESS` is garanteed to trigger when the following conditions are met:
1. jit is enabled
2. Runs on a slightly smaller batch size (e.g., `batch_size - 1`) compared to initialization (which triggers a jit re-compilation)
3. When batch size is greater than 16

I've tested `batch_size=8` and it works like a charm.

## Reproduction Command Line
```
# This crashes
$ python3 debug.py --batch-size 16 --jit

# The following does not crash
$ python3 debug.py --batch-size 16 
$ python3 debug.py --batch-size 8 --jit
$ python3 debug.py --batch-size 8 
```

I further tested the following variants and **they all run well**:
1. Fewer stages in network
    - change `stage_sizes = [2, 2, 3, 2]` to stage_sizes = [2, 2, 3]`
2. Change resblock from bottleneck structure to a single convolution
     - make `def __call__(self, x):` just returns `nn.Conv(c // 4, (3, 3), (s, s))(x)`

## Crash Logs
```
$ python3 debug.py --batch-size 16 --jit
2021-09-04 21:00:52.830628: E external/org_tensorflow/tensorflow/stream_executor/cuda/cuda_driver.cc:1155] failed to enqueue async memcpy from device to host: CUDA_ERROR_ILLEGAL_ADDRESS: an illegal memory access was encountered; host dst: 0x3f8de740; GPU src: 0x7f3bfabf7d00; size: 64=0x40
2021-09-04 21:00:52.830749: E external/org_tensorflow/tensorflow/stream_executor/stream.cc:334] Error recording event in stream: Error recording CUDA event: CUDA_ERROR_ILLEGAL_ADDRESS: an illegal memory access was encountered; not marking stream as bad, as the Event object may be at fault. Monitor for further errors.
2021-09-04 21:00:52.830790: E external/org_tensorflow/tensorflow/stream_executor/cuda/cuda_driver.cc:611] unable to add host callback: CUDA_ERROR_ILLEGAL_ADDRESS: an illegal memory access was encountered
Traceback (most recent call last):
  File "debug.py", line 130, in <module>
    main()
  File "debug.py", line 124, in main
    print(state.params['conv_init']['bias'])
  File "/home/zxy/.local/lib/python3.6/site-packages/jax/interpreters/xla.py", line 1060, in _forward_method
    return fun(getattr(self, attrname), *args)
RuntimeError: Internal: stream did not block host until done; was already in an error state
2021-09-04 21:00:52.967591: E external/org_tensorflow/tensorflow/stream_executor/cuda/cuda_driver.cc:1039] could not synchronize on CUDA context: CUDA_ERROR_ILLEGAL_ADDRESS: an illegal memory access was encountered :: *** Begin stack trace ***


















        PyDict_SetItem
        _PyModule_ClearDict
        PyImport_Cleanup
        Py_FinalizeEx
        Py_Main
        main
        __libc_start_main
        _start
*** End stack trace ***

2021-09-04 21:00:52.967745: F external/org_tensorflow/tensorflow/compiler/xla/service/gpu/gpu_executable.cc:89] Check failed: pair.first->SynchronizeAllActivity()
[1]    5164 abort      python3 debug.py --batch-size 16 --jit
```
## Non-crash Log
```
$ python3 debug.py --batch-size 8 --jit                                                                        130
[-0.00078839 -0.00099954 -0.00099411  0.00100002  0.00100001 -0.00083913
 -0.00099661 -0.00100002  0.00099884 -0.00100002 -0.00099947 -0.00100002
  0.00100002 -0.00098644  0.00099998  0.0009997 ]
Have a nice day!
```

## System Info
- Ubuntu 18.04
- nvidia 460.32.03
- cuda 10.2
- cudnn 7.6.5
- python packages
     - flax==0.3.4
     - jax==0.2.17
     - jaxlib==0.1.68+cuda102
     - optax==0.0.9

**Comment by zxytim** [comment]
If I change `GroupNorm` to `LayerNorm` (the one in the `flax.linen` package), `python3 debug.py --batch-size 16 --jit` will exit gracefully as well.

**Comment by hawkinsp** [comment]
There are a number of known `CUDA_ERROR_ILLEGAL_ADDRESS` bugs for CUDA < 11.1 that are out of our control (they are bugs in `ptxas`). Can you upgrade to a CUDA >= 11.1? I strongly suspect that will fix your bug.

**Comment by hawkinsp** [comment]
(I should add: we intend to drop CUDA 10.2 support in the next release and only support 11.1+ for precisely this reason.)

**Comment by hawkinsp** [comment]
Closing, because I'm fairly confident this is a CUDA problem that would be fixed by upgrading CUDA to 11.1+. Please reopen if the problem still occurs with an up to date CUDA.
